pallets/stateful-storage/src/lib.rs

```diff
@@ -0,0 +1,797 @@
+		SchemaPayloadLocationMismatch,
+
+		/// Invalid Message Source Account
+		InvalidMessageSourceAccount,
+
+		/// UnauthorizedDelegate
+		UnauthorizedDelegate,
+
+		/// Corrupted State
+		CorruptedState,
+
+		/// Invalid item action
+		InvalidItemAction,
+
+		/// Target page hash does not match current page hash
+		StalePageState,
+
+		/// Invalid Signature for payload
+		InvalidSignature,
+
+		/// The submitted proof has expired; the current block is less the expiration block
+		ProofHasExpired,
+
+		/// The submitted proof expiration block is too far in the future
+		ProofNotYetValid,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An event for when an itemized storage is updated
+		ItemizedPageUpdated {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// previous content hash before update
+			prev_content_hash: PageHash,
+			/// current content hash after update
+			curr_content_hash: PageHash,
+		},
+
+		/// An event for when an itemized storage is deleted
+		ItemizedPageDeleted {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// previous content hash before removal
+			prev_content_hash: PageHash,
+		},
+
+		/// An event for when an paginated storage is updated
+		PaginatedPageUpdated {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// id of updated page
+			page_id: PageId,
+			/// previous content hash before update
+			prev_content_hash: PageHash,
+			/// current content hash after update
+			curr_content_hash: PageHash,
+		},
+
+		/// An event for when an paginated storage is deleted
+		PaginatedPageDeleted {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// id of updated page
+			page_id: PageId,
+			/// previous content hash before removal
+			prev_content_hash: PageHash,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Applies the Add or Delete Actions on the requested Itemized page
+		///
+		/// # Events
+		/// * [`Event::ItemizedPageUpdated`]
+		/// * [`Event::ItemizedPageDeleted`]
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::apply_item_actions( actions.len() as u32 ,
+			actions.iter().fold(0, |acc, a| acc + match a {
+				ItemAction::Add { data } => data.len() as u32,
+				_ => 0,
+			}),
+		    0
+		))]
+		pub fn apply_item_actions(
+			origin: OriginFor<T>,
+			#[pallet::compact] state_owner_msa_id: MessageSourceId,
+			#[pallet::compact] schema_id: SchemaId,
+			#[pallet::compact] target_hash: PageHash,
+			actions: BoundedVec<ItemAction, T::MaxItemizedActionsCount>,
+		) -> DispatchResult {
+			let provider_key = ensure_signed(origin)?;
```

Comment:
Maybe we do not call this provider_key and instead just key. This makes me think that the signer needs to be a provider.